crates/bevy_math/src/curve/mod.rs

```diff
@@ -212,12 +215,194 @@ pub trait Curve<T> {
         })
     }
 
+    /// Resample this [`Curve`] to produce a new one that is defined by interpolation over equally
+    /// spaced sample values, using the provided `interpolation` to interpolate between adjacent samples.
+    /// The curve is interpolated on `segments` segments between samples. For example, if `segments` is 1,
+    /// only the start and end points of the curve are used as samples; if `segments` is 2, a sample at
+    /// the midpoint is taken as well, and so on. If `segments` is zero, or if this curve has an unbounded
+    /// domain, then a [`ResamplingError`] is returned.
+    ///
+    /// The interpolation takes two values by reference together with a scalar parameter and
+    /// produces an owned value. The expectation is that `interpolation(&x, &y, 0.0)` and
+    /// `interpolation(&x, &y, 1.0)` are equivalent to `x` and `y` respectively.
+    ///
+    /// # Example
+    /// ```
+    /// # use bevy_math::*;
+    /// # use bevy_math::curve::*;
+    /// let quarter_rotation = function_curve(interval(0.0, 90.0).unwrap(), |t| Rot2::degrees(t));
+    /// // A curve which only stores three data points and uses `nlerp` to interpolate them:
+    /// let resampled_rotation = quarter_rotation.resample(3, |x, y, t| x.nlerp(*y, t));
+    /// ```
+    fn resample<I>(
+        &self,
+        segments: usize,
+        interpolation: I,
+    ) -> Result<SampleCurve<T, I>, ResamplingError>
+    where
+        Self: Sized,
+        I: Fn(&T, &T, f32) -> T,
+    {
+        if segments == 0 {
+            return Err(ResamplingError::NotEnoughSamples(segments));
+        }
+        if !self.domain().is_bounded() {
+            return Err(ResamplingError::UnboundedDomain);
+        }
+
+        let samples: Vec<T> = self
+            .domain()
+            .spaced_points(segments + 1)
+            .unwrap()
+            .map(|t| self.sample_unchecked(t))
+            .collect();
```

Comment:
Btw: Can't we just use `samples` instead of `segments` as an argument here as well to stay consistent with the next function or does it make more sense like this semantically?

```suggestion
        let samples: Vec<T> = self.samples(segments+1)?.collect();
```

Reply:
I found `samples` to be kind of confusing when using this myself, since I usually think more directly of, e.g., breaking a curve into two segments, rather than sampling it three times to make those segments.

Reply:
Well, that's ok. It shouldn't be a blocker. Let's see how it turns out when we actually use it in the end :+1: